### ISSUE TITLE
Different results when vocoder model is loaded

### ISSUE THREAD

**Issue by FaridHaziyev** [espnet/espnet#2946]
Below code snippet is from E2E-TTS demo colab. When I load the vocoder, then results of the tacotron model changes. I think this may be something about pytorch, but I tried it in a different scenario. I do not use vocoder but use grifin lim algorithm for mel2wav, and test what wil happen if I load the vocoder model or not. When I just load the vocoder model but do not use it, results changed, compared to not loading the vocoder. 

```
model_class = dynamic_import(train_args.model_module)
model = model_class(idim, odim, train_args)
torch_load(model_path, model)
model = model.eval().to(device)
inference_args = Namespace(**{
    "threshold": 0.5,"minlenratio": 0.0, "maxlenratio": 10.0,
    # Only for Tacotron 2
    "use_attention_constraint": True, "backward_window": 1,"forward_window":3,
    # Only for fastspeech (lower than 1.0 is faster speech, higher than 1.0 is slower speech)
    "fastspeech_alpha": 1.0,
    })

# define neural vocoder
from parallel_wavegan.utils import load_model
fs = 22050
vocoder = load_model(vocoder_path)
vocoder.remove_weight_norm()
vocoder = vocoder.eval().to(device)

```

**Comment by kan-bayashi** [comment]
See
https://github.com/espnet/espnet/issues/2647
https://github.com/espnet/espnet/issues/2875

**Comment by FaridHaziyev** [comment]
Actually I set the manual seed, and it gives same results each time but what I want to say is when I load the vocoder model, it changes the seed or something else happens and results changes. I do not use vocoder model, I just load it and it changes the result of the tacotron model. 

In another scenario I do the char2mel at first then load the vocoder model and it gives the best result. Then I load tacotron model first then load the vocoder and then set manual seed. This way result is the same each time, but not as good as the first scenario. So I believe other configs are also changed by loading vocoder model.

**Comment by FaridHaziyev** [comment]
What I see is that when I load vocoder model it changes the tacotron model

**Comment by kan-bayashi** [comment]
> In another scenario I do the char2mel at first then load the vocoder model and it gives the best result. Then I load tacotron model first then load the vocoder and then set manual seed. This way result is the same each time, but not as good as the first scenario.

I could not understand the meaning of `but not as good as the first scenario`.
The latter one seems the best way to fix the seed.

**Comment by FaridHaziyev** [comment]
First scenario is I apply tacotron model then load vocoder and apply vocoder model.

**Comment by FaridHaziyev** [comment]
I want to find what configuration is changed when I load the vocoder model

**Comment by kan-bayashi** [comment]
Why do you need to load the model every time?
I could not imagine. Why don't you fix the seed again?

**Comment by FaridHaziyev** [comment]
Even when I fix the seed again, results are not the same, they are changed.

**Comment by FaridHaziyev** [comment]
Loading again means, load tacotron apply it then load vocoder and apply

**Comment by kan-bayashi** [comment]
Please share the code to reproduce.

**Comment by kan-bayashi** [comment]
The randomness came from
- Dropout of Prenet in Tacotron2
- Input noise if you use PWG

**Comment by FaridHaziyev** [comment]
first one is like this :

`def char2mel(texts):
    set_deterministic_pytorch(Namespace(**{"seed":1, "debugmode":1}))
    idim, odim, train_args = get_model_conf(model_path)
    model_class = dynamic_import(train_args.model_module)
    model = model_class(idim, odim, train_args)
    torch_load(model_path, model)
    model = model.eval().to(device)

    mels = []
    with torch.no_grad():
        for text in texts:
            tokenized_text = tokenize(text)
            input_data = torch.LongTensor(tokenized_text).to(device)
            c_, _, _ = taco_model.inference(input_data, inference_args)
            mels.append(c_)
    return mels


def mel2wav(mels):
    audios = []
    vocoder = load_model(vocoder_path)
    vocoder.remove_weight_norm()
    vocoder = vocoder.eval().to(device)

    with torch.no_grad():
        for mel in mels:
            y_ = vocoder_model.inference(mel)
            y = y_.view(-1).cpu().data.numpy()
            audios.append(y)
    audios = np.concatenate(audios)
    write("testing_vocoder.wav", 22050, (audios * np.iinfo(np.int16).max).astype(np.int16))
`

**Comment by FaridHaziyev** [comment]
second is like :
```
vocoder = load_model(vocoder_path)
vocoder.remove_weight_norm()
vocoder = vocoder.eval().to(device)

set_deterministic_pytorch(Namespace(**{"seed":1, "debugmode":1}))


idim, odim, train_args = get_model_conf(model_path)
model_class = dynamic_import(train_args.model_module)
model = model_class(idim, odim, train_args)
torch_load(model_path, model)
model = model.eval().to(device)




def char2wav(texts):
    audios = []
    with torch.no_grad():
        for text in texts:
            tokenized_text = tokenize(text)
            input_data = torch.LongTensor(tokenized_text).to(device)
            mel, _, _ = taco_model.inference(input_data, inference_args)
            y_ = vocoder_model.inference(mel)
            y = y_.view(-1).cpu().data.numpy()
            audios.append(y)
```

**Comment by kan-bayashi** [comment]
Please write the code to reproduce the error.

e.g.
```python
# load text2mel
idim, odim, train_args = get_model_conf(model_path)
model_class = dynamic_import(train_args.model_module)
model = model_class(idim, odim, train_args)
torch_load(model_path, model)
model = model.eval().to(device)

# load vococder
vocoder = load_model(vocoder_path)
vocoder.remove_weight_norm()
vocoder = vocoder.eval().to(device)

# setup input
text = "this is an example.
tokenized_text = tokenize(text)
input_data = torch.LongTensor(tokenized_text).to(device)

# first inference
set_deterministic_pytorch(Namespace(**{"seed":1, "debugmode":1}))
mel, _, _ = taco_model.inference(input_data, inference_args)
y_ = vocoder_model.inference(mel)
y1 = y_.view(-1).cpu().numpy()

# second inference
set_deterministic_pytorch(Namespace(**{"seed":1, "debugmode":1}))
mel, _, _ = taco_model.inference(input_data, inference_args)
y_ = vocoder_model.inference(mel)
y2 = y_.view(-1).cpu().numpy()

# assert
np.testing.assert_array_equal(y1, y2)

```

**Comment by kan-bayashi** [comment]
I could not reproduce.
https://gist.github.com/kan-bayashi/0a4911719ca04e73f378b45c335c2dc4

**Comment by FaridHaziyev** [comment]
```
idim, odim, train_args = get_model_conf(model_path)
model_class = dynamic_import(train_args.model_module)
model = model_class(idim, odim, train_args)
torch_load(model_path, model)
model = model.eval().to(device)

# load vococder
vocoder = load_model(vocoder_path)
vocoder.remove_weight_norm()
vocoder = vocoder.eval().to(device)

# setup input
text = "mene deniz verin"
tokenized_text = tokenize(text)
input_data = torch.LongTensor(tokenized_text).to(device)

# first inference
with torch.no_grad():
    set_deterministic_pytorch(Namespace(**{"seed":1, "debugmode":1}))
    mel1, _, _ = model.inference(input_data, inference_args)
    #y_ = vocoder.inference(mel1)
    #y1 = y_.view(-1).cpu().numpy()


set_deterministic_pytorch(Namespace(**{"seed":1, "debugmode":1}))
idim, odim, train_args = get_model_conf(model_path)
model_class = dynamic_import(train_args.model_module)
model = model_class(idim, odim, train_args)
torch_load(model_path, model)
model = model.eval().to(device)

# setup input
text = "mene deniz verin"
tokenized_text = tokenize(text)
input_data = torch.LongTensor(tokenized_text).to(device)


# second inference
with torch.no_grad():
    mel2, _, _ = model.inference(input_data, inference_args)

# assert
print(np.testing.assert_array_equal(mel1, mel2))

```

**Comment by FaridHaziyev** [comment]
I think my problem was this. This way assertion fails

**Comment by kan-bayashi** [comment]
This seems reasonable because you define the model after setting seed.
When initializing the model, the random initialization will be performed.
Therefore, the seed will be updated.
You need to fix the seed just before inference.

**Comment by FaridHaziyev** [comment]
thank you very much
